### PR TITLE
fix(ui): close floating window on BufLeave event

### PR DESCRIPTION
**Pull request by ii14**
When buffer is visible in two splits simultaneously, BufHidden event is
not triggered, causing the floating window to remain on screen after
switching to another buffer.

Remove BufHidden event from close_events defaults, and close the window
if we changed the buffer to something other than the buffer that spawned
the floating window or the floating window buffer itself.

### CODE REVIEW

**Review by gpanders — DISMISSED** [review]

**Review by mfussenegger — COMMENTED** [review]
I think this breaks entering the window. E.g. try `vim.lsp.buf.hover()` followed by `CTRL-W w`

**Comment by ii14** [comment]
This is kinda problematic, mainly because events are configurable by user and this requires special handling. Solving this might require some changes to the API. Or is vim.lsp.util considered private and I shouldn't be bothered by this too much?

**Comment by ii14** [comment]
I managed to do it without changing `vim.lsp.util.open_floating_preview` and it works, but because of the special handling that is necessary, I'm replacing `BufLeave` event with `BufEnter`, and I'm not sure if this is the best way to go about it. I think the behavior of closing the floating window on leaving the buffer could be a separate option, on top of close_events.

Also autocmds were leaking, so I fixed it with augroups.

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -1224,13 +1224,59 @@ end
 ---
 ---@param events (table) list of events
 ---@param winnr (number) window id of preview window
+---@param bufnrs (table) buffers where the preview window should remain visible
 ---@see |autocmd-events|
-function M.close_preview_autocmd(events, winnr)
-  if #events > 0 then
-    api.nvim_command("autocmd "..table.concat(events, ',').." <buffer> ++once lua pcall(vim.api.nvim_win_close, "..winnr..", true)")
+function M.close_preview_autocmd(events, winnr, bufnrs)
```
Comment (gpanders) [inline runtime/lua/vim/lsp/util.lua:1229]: Does this need to be a public function? grep shows it only used in util.lua, and somehow I strongly doubt plugins are using it.

**Review by gpanders — COMMENTED** [review]
Can you add some comments explaining the motivation and the (high-level) implementation? I imagine future readers not named ii14 will wonder why this exists the way it does.

**Comment by gpanders** [comment]
@ii14 and I discussed on Matrix making the behavior of "close the preview window when the buffer that opened it closes" the default and non-configurable. This seems like a sane default to me (and we can always put it behind an option one day if necessary), but if anyone feels differently drop a comment.

**Comment by ckipp01** [comment]
By chance is there a recommendation for existing plugin authors that make use of this what they should use as an alternative?

**Comment by ii14** [comment]
You mean for close_preview_autocmd function or the floating window in general?

For floating window it's only a breaking change if you are using custom close_events.

As for close_preview_autocmd function, it was literally just 5 lines of code:
```lua
function M.close_preview_autocmd(events, winnr)
  if #events > 0 then
    api.nvim_command("autocmd "..table.concat(events, ',').." <buffer> ++once lua pcall(vim.api.nvim_win_close, "..winnr..", true)")
  end
end
```

**Comment by ii14** [comment]
close_preview_autocmd was already called in open_floating_preview for you automatically, that's why there was no real reason for it to be public, and it's such a short function.

You can still add more custom events to close_events, however currently there is no option to disable the behavior of closing the floating window on leaving the buffer, but I can add it if you need it.

**Comment by ckipp01** [comment]
> You can still add more custom events to close_events, however currently there is no option to disable the behavior of closing the floating window on leaving the buffer, but I can add it if you need it.

Nah, I don't think so. I'll just look more into exactly what was happening and mimic it if needed.

**Comment by solarizedalias** [comment]
Hi, I think `close_preview_autocmd()` had already been used by plugins.
Not sure if exporting it again makes sense here but just for your information.

https://github.com/neovim/nvim-lspconfig/blob/aa95809aa5a518c5979df7ebfcfb63a2528ea4ad/lua/lspconfig/ui/lspinfo.lua#L206
https://github.com/tami5/lspsaga.nvim/blob/54b76c15f9a602b6b7110b153335ff7e0f690e7a/lua/lspsaga/diagnostic.lua#L98

**Comment by mg979** [comment]
@ii14 Just updated to 0.6.1 and saw `close_preview_autocmd` gone and throwing errors. It was handy to set up autocommands to close the window even when not using `open_floating_preview`, just `nvim_open_win`. Is there another way to do that now? That isn't those _5 lines of codes_ you mentioned.

I mean it's terribly off-putting to find out breaking changes that aren't even documented in the release notes. Sorry for the rant.

**Comment by ii14** [comment]
@mg979 you can look at the diff, that's what this function used to be and it should work exactly as documented, unless there was some other changes made before this commit that I'm not aware of. I'm just guessing, but you need to somehow expose this function, eg. return it from a module, and maybe you didn't do that? For example you could change `M.close_preview_autocmd` to `_G.close_preview_autocmd` to make it a global variable, and then replace calls from `vim.lsp.util.close_preview_autocmd(...)` to just `close_preview_autocmd(...)`.

Sorry, I'm not a maintainer and I have no idea how releases are made, so I can't tell you anything about that. But it indeed does look like breaking changes are missing from 0.6.1 release notes.

@gpanders

**Comment by mg979** [comment]
@ii14 yes I copied the old function (as it was before this PR) where I needed it and it still works. I just don't think it's been a good idea to remove it, because API for popups is already difficult and this helped. Maybe a bit strange that it's buried in some `lsp.util` module, but it helped. At least I was using it outside of lsp stuff.

**Comment by Julian** [comment]
>  currently there is no option to disable the behavior of closing the floating window on leaving the buffer, but I can add it if you need it.

We rely on this I believe (the old behavior) -- we open groups of floating windows (tooltips) that are intended to stay open as nested ones are open, which now we can't configure to stay open.

**Pull request merged**